ee/server/service/mdm.go

```diff
@@ -742,7 +745,109 @@ func (svc *Service) MDMAppleMatchPreassignment(ctx context.Context, externalHost
+	if err != nil || !hostMDM.IsFleetEnrolled() {
+		if err == nil || fleet.IsNotFound(err) {
+			err = errors.New("host is not enrolled in Fleet MDM")
+			return ctxerr.Wrap(ctx, &fleet.BadRequestError{
+				Message:     err.Error(),
+				InternalErr: err,
+			})
+		}
+		return err
+	}
+
+	// Collect the profiles' hashes and look for a team with exactly that set.
+	// Also collect the profiles' groups in case we need to create a new team,
+	// and the list of raw profiles bytes.
+	hashes, groups, rawProfiles := make([]string, 0, len(profs.Profiles)),
+		make([]string, 0, len(profs.Profiles)),
+		make([][]byte, 0, len(profs.Profiles))
+	for _, prof := range profs.Profiles {
+		hashes = append(hashes, prof.HexMD5Hash)
+		rawProfiles = append(rawProfiles, prof.Profile)
+		if prof.Group != "" {
+			groups = append(groups, prof.Group)
+		}
+	}
+
+	// find a team with exactly that set of profiles
+	teamIDs, err := svc.ds.MatchMDMAppleConfigProfiles(ctx, hashes)
+	if err != nil {
+		return err
+	}
+
+	var targetTeamID uint
+	if len(teamIDs) > 0 {
+		// if the host is already in one of those valid teams, nothing to do.
+		if host.TeamID != nil {
+			for _, tmID := range teamIDs {
+				if *host.TeamID == tmID {
+					return nil
+				}
+			}
+		}
+		// else assign the host to the first valid team
+		targetTeamID = teamIDs[0]
+
+	} else {
+		// create a new team with this set of profiles
+		dedupeGroups := make(map[string]struct{}, len(groups))
+		for _, group := range groups {
+			dedupeGroups[group] = struct{}{}
+		}
+		groups = groups[:0]
+		for group := range dedupeGroups {
+			groups = append(groups, group)
+		}
+		sort.Strings(groups)
+
+		if len(groups) == 0 {
+			groups = []string{"default"}
+		}
+
+		// creating via the service call so that it properly assigns the agent
+		// options and creates audit activities, etc.
+		teamName := fmt.Sprintf("%s (%s)", strings.Join(groups, " - "), time.Now().UTC().Format("2006-01-02:15:04:05"))
+		tm, err := svc.NewTeam(ctx, fleet.TeamPayload{Name: &teamName})
+		if err != nil {
+			return err
+		}
+
+		// create profiles for that team via the service call, so that uniqueness
+		// of profile identifier/name is verified, activity created, etc.
+		// NOTE: this will use the read replica to load the team, which was just
+		// created above, could lead to not found issues with slow replication.
+		if err := svc.BatchSetMDMAppleProfiles(ctx, &tm.ID, nil, rawProfiles, false); err != nil {
+			return err
+		}
+
+		targetTeamID = tm.ID
+	}
+
+	// assign host to that team via the service call, which will trigger
+	// deployment of the profiles.
+	if err := svc.AddHostsToTeam(ctx, &targetTeamID, []uint{host.ID}); err != nil {
+		return err
+	}
+
+	// TODO(mna): should that create an audit activity? Note that the
+	// "edited_macos_profiles" and "team_created" activities will be created if
+	// the team was created, but I'm thinking "puppet-assign host" activity or
+	// something.
```

Comment:
Question for @noahtalerman , though this can go in a future PR if needed, no need to block this one for that.

Reply:
Noah is out for ~2 weeks, cc: @zhumo 

My vote is to default to creating a ticket to follow up, regardless of the decision.

Reply:
Ticket created: https://github.com/fleetdm/fleet/issues/12129